firestore/src/main/java/com/firebase/ui/firestore/paging/FirestoreItemKeyedDataSource.kt

```diff
@@ -0,0 +1,263 @@
+     * Tries to load half of [loadSize] documents before [document] and half of [loadSize] documents starting at provided [document].
+     *
+     * @param document The document to load around.
+     * @param loadSize Requested number of items to load.
+     *
+     * @return A Task that will be resolved with the results of the Query.
+     */
+    private fun loadAround(
+            document: DocumentSnapshot,
+            loadSize: Long,
+            callback: LoadInitialCallback<DocumentSnapshot>
+    ) = tryGetDocumentPrior(document, loadSize / 2).continueWithTask { previousTask ->
+        loadStartAt(previousTask.result!!, loadSize, callback)
+    }
+
+    /**
+     * Returns document before [document] maximum by [preloadSize] or [document] if there is no documents before.
+     *
+     * @param document The document to load before.
+     * @param preloadSize Requested number of items to preload.
+     *
+     * @return A Task that will be resolved with the results of the Query.
+     */
+    private fun tryGetDocumentPrior(
+            document: DocumentSnapshot,
+            preloadSize: Long
+    ): Task<DocumentSnapshot> = getItemsBefore(document, preloadSize).continueWith { task ->
+        if (task.isSuccessful) {
+            task.result!!.documents.lastOrNull() ?: document
+        } else {
+            document
+        }
+    }
+
+    private fun getItemsBefore(
+            document: DocumentSnapshot,
+            loadSize: Long
+    ) = queryEndAt(document, loadSize).get(CACHE)
+
+    private fun loadStartAt(
+            document: DocumentSnapshot,
+            loadSize: Long,
+            callback: LoadInitialCallback<DocumentSnapshot>
+    ) = PageLoader(callback).also { pageLoader ->
+        addPageLoader(queryStartAt(document, loadSize), pageLoader)
+    }.task
+
+    private fun loadFromBeginning(
+            loadSize: Long,
+            callback: LoadInitialCallback<DocumentSnapshot>
+    ) {
+        addPageLoader(
+                queryStartFromBeginning(loadSize),
+                PageLoader(callback)
+        )
+    }
+
+    override fun loadAfter(
+            params: LoadParams<DocumentSnapshot>,
+            callback: LoadCallback<DocumentSnapshot>
+    ) {
+        addPageLoader(
+                queryStartAfter(params.key, params.requestedLoadSize.toLong()),
+                PageLoader(callback)
+        )
+    }
+
+    override fun loadBefore(
+            params: LoadParams<DocumentSnapshot>,
+            callback: LoadCallback<DocumentSnapshot>
+    ) {
+        addPageLoader(
+                queryEndBefore(params.key, params.requestedLoadSize.toLong()),
+                PageLoader(callback.reversedResults())
+        )
+    }
+
+    override fun getKey(item: DocumentSnapshot) = item
+
+    override fun invalidate() {
+        super.invalidate()
+        removeAllPageLoaders()
+    }
+
+    private fun removeAllPageLoaders() {
+        registrations.forEach { it.remove() }
+        registrations.clear()
```

Comment:
I believe one of the reasons this hasn't been done yet is because of how difficult it is to properly subscribe and unsub from these listeners. Can you explain a bit more how and when `invalidate` is called?

Reply:
No, I just forget to unsubscribe.

I wrote it previously. By design of Paging library to refresh data in UI you should invalidate DS.
Here quote from Paging library documentation:

> A PagedList / DataSource pair are a snapshot of the data set. A new pair of PagedList / DataSource must be created if an update occurs, such as a reorder, insert, delete, or content update occurs. A DataSource must detect that it cannot continue loading its snapshot (for instance, when Database query notices a table being invalidated), and call invalidate(). Then a new PagedList / DataSource pair would be created to load data from the new state of the Database query.

Reply:
Ok, I think I understand what's going on now. 👍

I'm a little concerned about cost. Correct me if I'm wrong, but here's what I think is going on: when a page is being loaded, it adds a snapshot listener which submits the data the first time it gets triggered. Then, on the second trigger, it invalidates the entire data set and removes all the listeners. This causes the adapter to requery everything from scratch and start the process anew. Is that correct?

If so, this feature would be extremely expensive to users. They wouldn't be paying for the diff—every single update would cost them the entire query. Coming from an official Firebase library, that seems a little abusive. @samtstern WDYT?

Reply:
Yes, you right that it requery data on every chnge, but not all the data only recent pageSize*3 items by default or can be specified by PagedList.Config.initialLoadSizeHint. What is more when you finish #1388 there will be PagedList.Config.maxSize which limits number of item currently kept in memory and which will be involved in old with new data comparison.

What about requerying this 3 resent pages most likely will be loaded from cache if I understand correctly how firebase works.

Reply:
After some testing I come to conclusion that I was wrong - every requery leeds to extra charges.
I was wrongly assumed that since offline access is enabled by default recently accessed documents will be fetched from local cache and do not count as priced reads. But unfortunately thats not true.

Reply:
If you issue the **exact same** query multiple times within about 30 minutes, you shouldn't be charged for the entire doc set multiple times.  You'll only be charged for the new docs.

Reply:
Yes, you right, but problem is that exactly same means with the same limit and start/end conditions.
But in this implementation we can't avoid creating new queries starting from different positions so I think it is expensive to use. 
I have an idea how to solve this issue. If I succeed I send new PR.

Reply:
Assume that the query result is identical

> execute exact same query multiple times doesn't charge for the entire doc set multiple times

Is this still true? @samtstern

Reply:
Yes that's true, identical queries issued within 30m of each other don't double-charge unless there are new results.

Reply:
What if we have 1 new document in the query result, for example, the first query return 3 documents, the second query return the same 3 documents + 1 new document. In this case, how many documents read will be charged?

Reply:
In that case 4 total reads.  3 for the initial result.  1 for the update later.